リファクタリング用空コミット

LakeRainSound opened the pull request:
気になる点．
- エラーハンドリング(get_code_metrics/code_metrics/cloc/cloc_analyze.py, get_code_metrics/github_api/label_info.py, get_code_metrics/github_api/repository_info.py)
- dictionaryの結果の統合の仕方(get_code_metrics/gcm_output/output.py, get_code_metrics/github_api/github_api.py _get_linked_github_api_resultメソッド)
- スレッドの使い方(スレッドとしてのエラーハンドリングが必要かなど)
- その他，不具合になりうる点，コードとして改善できる点．

pddg commented:
## 処理内容で気になった点

- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/gcm_output/output.py#L28-L31 と https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/github_api.py#L13-L17
    - ここやりたいことがよく分からない．`dict_list`の各要素には何が入っている予定なの？
- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/label_info.py#L96-L103
    - ここかなりよく分からない

## エラー系で気になった点

- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/code_metrics/cloc/cloc_analyze.py#L24-L30
    - 単にerrorという文字列を含むかどうかなら`'error' in cloc_result`で良い
    - ほんとにこれでエラー検知は上手くいっているの？
- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/post_query.py#L5-L13
    - ここ，Exceptionをraiseする割にはこれを呼び出す箇所のどこでもexceptする記述が無くない？
    - ここでエラーが起きるとスクリプト全体が止まる気がする

## おそらく影響は無いが気になった点

- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/code_metrics/cloc/cloc_analyze.py#L8-L13 と https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/post_query.py#L6-L9
    - インスタンス変数と名前が重複しているクラス変数の謎
- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/code_metrics/cloc/cloc_analyze.py#L47
    - Pathオブジェクトは演算子オーバーライドによってパス区切りを`/`で書けるので，`self.path_to_ghq_root / 'github.com' / repository_name`で良い
- RateLimitを見てSleepする処理，クエリ投げるところ全部で書いているので https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/post_query.py の中で記述してしまっても良いと思う
- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/cli.py#L59
    - 改行コード削除は`repository = repository.rstrip(os.linesep)`
- https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/cli.py#L61
    - ここ意味一緒
- `import ~~~.~~~~.~~.~~~ as ~~`を使うよりは`from ~~~.~~~~.~~ import ~~~`の方が良いと思う

## 全体的な感想

- ところどころPythonぽくないけど，責務が分かれているので読みやすい
- READMEに最終的にどういうjsonが得られるのか書いておくとわかりやすいと思う
- 使用方法についてもREADMEに記述が欲しい
- 全体のProgress知りたい…知りたくない？tqdm使おう？（ラボWikiに使い方書いてます）

LakeRainSound commented:
@pddg
長くなりそうなので，まずは**処理内容で気になった点**の部分から説明とやりたいことを書いていきます．https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/github_api.py#L13-L17 で指摘を受けた部分ですが，まずresult_listには，
```
{
    "repository": {
        "LakeRainSound/get_code_metrics": {
            "nameWithOwner": "LakeRainSound/get_code_metrics",
            "createdAt": "2019-12-18T11:39:49Z",
            "stargazers": {
                "totalCount": 1
            },
            "hasIssuesEnabled": true,
            "isArchived": false,
            "isFork": false,
            "isDisabled": false,
            "url": "https://github.com/LakeRainSound/get_code_metrics"
        }
    }
}
```
と
```
{
    "repository": {
        "LakeRainSound/get_code_metrics": {
            "closedIssueCount": 5,
            "hasLabelClosedIssue": 5
        }
    }
}
```
の2つのdictが入っています(今回は例示のため，リポジトリは一つにしていますが本来は，複数のリポジトリがそれぞれでならんでいると思っていただきたいです)．
ここで一つ目のjsonを結果を返す`github_results`に入れています．
```
github_results.update({'repository': result_list[0]})
```
このあと僕が上手い書き方がわからなかったので，指摘されている部分だと思うのですが，やりたかったこととしては，`repository`の中で各リポジトリ(name with ownerがキーになっている)の情報があるので，その各要素(リポジトリ)を1つ目にはなかった要素をもつ2つめのjson(closedIssueCountとhasLabelClosedIssueを含んでいる方)に追加したかったということです．
**ここは調べてもちゃんと出てこなかったので楽かつわかりやすい書き方があれば教えていただきたいです．**
https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/gcm_output/output.py#L28-L31 も同様です．

もう一つの https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/github_api/label_info.py#L96-L103 について述べていきます．
まず`issues`に入っているjsonの形式ですが，
```
{
    "title_and_label": [
        {
            "title": "\u30b3\u30de\u30f3\u30c9\u30e9\u30a4\u30f3\u5f15\u6570\u306e\u30d1\u30fc\u30b9",
            "labels": {
                "labelCount": 1
            }
        },
        {
            "title": "\u540c\u3058\u30c7\u30a3\u30ec\u30af\u30c8\u30ea\u306b\u3042\u308b.py\u30d5\u30a1\u30a4\u30eb\u304cimport\u3067\u304d\u306a\u3044",
            "labels": {
                "labelCount": 1
            }
        },
        {
            "title": "\u30d7\u30ed\u30b8\u30a7\u30af\u30c8\u5168\u4f53\u306e\u69cb\u6210",
            "labels": {
                "labelCount": 1
            }
        },
        {
            "title": "GitHub API\u3092\u53e9\u3044\u3066\u7d50\u679c\u3092\u8fd4\u3059\uff0e",
            "labels": {
                "labelCount": 1
            }
        },
        {
            "title": "cloc\u306e\u5b9f\u88c5\uff0e",
            "labels": {
                "labelCount": 2
            }
        }
    ],
    "closedIssueCount": 5
}
```
のようになっています(titleはissueのタイトルです．日本語なので文字コードに変換されているのでわかりづらくてすみません)．つまり，"title_and_label"の中に全てのissueオブジェクトがあり，その中にissue titleとlabelの種類数が格納されています(この形式はGitHub APIの仕様なので変更はできません)．
やりたいこととしては，一つのissueに対してラベルが付与されているかどうかを計上することなので，`labelCount`が >0 であるなら，has_label_countをインクリメントして最後に結果を返すためのdict変数`label_metrics`に`"hasLabelClosedIssue"`をキーとする値を
```
label_metrics.update({"hasLabelClosedIssue": has_label_count})
```
で入れるという形にしています．
この部分でわかりやすく記述を変えることができるなら教えていただきたいです．

LakeRainSound commented:
@pddg 
### エラー検知について
- clocの部分
>単にerrorという文字列を含むかどうかなら'error' in cloc_resultで良い
これは部分文字列に対して有効ってことでいいですか？
さらにそもそもについてですが，エラーが存在した場合のclocのエラーメッセージは.stdout.readlines()取得できて以下のようになっています．
```
1 error:

Unable to read:  /Users/restartsugar/.cache/gcm/repositories/github.com/lslakaflkjc/vhjksfhkavvn
```
そのため`error`というワードを調べる形で検査しています．
まずい場合は変更するべき点を教えていただきたいです．

- postの部分
確かに，200以外が返ってきた時点で止まってしまいますね...こういうのってどういう風(条件)にexceptを記述するのがいんでしょうか？もしくはそもそも200以外が返ってきても今回は無視した方がいいんでしょうか？

LakeRainSound commented:
@pddg 
その他の気になった部分に関しては修正しました．確かにと思える指摘が多くて見やすくなった箇所が多かったです．ありがとうございます．

pddg commented:
@LakeRainSound 

> そのためerrorというワードを調べる形で検査しています．
> まずい場合は変更するべき点を教えていただきたいです．

これで気になっているのは全行forで回してerrorという文字列を見つけようとしている点で，正常な結果を得た場合に標準出力にerrorという文字列が入らないという保証はあるのかな，と思いました．

pddg commented:
@LakeRainSound 

> 確かに，200以外が返ってきた時点で止まってしまいますね...こういうのってどういう風(条件)にexceptを記述するのがいんでしょうか？もしくはそもそも200以外が返ってきても今回は無視した方がいいんでしょうか？

基本的にHTTPリクエストは失敗する可能性が高いので，50x系が来たらとりあえず何度か再送します．その上で失敗したら，「そのリクエストが失敗した時に破棄して良いタスクの範囲」でexceptして大域脱出します．例えば，forループでリポジトリのリストを回して各リポジトリの結果を取得する際，一つ失敗したら全部やめるという場合はforの外側で，1つ失敗しても続けるならその内側でexceptして何かエラー出力を残します．どういう原因で失敗したのかを知るために，そのエラーメッセージも一緒に保存することを推奨します．

pddg commented:
@LakeRainSound 

> やりたかったこととしては，repositoryの中で各リポジトリ(name with ownerがキーになっている)の情報があるので，その各要素(リポジトリ)を1つ目にはなかった要素をもつ2つめのjson(closedIssueCountとhasLabelClosedIssueを含んでいる方)に追加したかったということです．

つまり，clocの結果と，GitHub APIからの結果の辞書をマージしたいってことだよね？これ今片方の辞書のキー（リポジトリの名前）だけ見てfor回しているけれど，片方はエラーでもう片方はエラーでない場合に取りこぼしが出そうなので，まずその和集合を取った方がいいと思います．

```python
r1 = dict_list[0]['repository'].keys()
r2 = dict_list[1]['repository'].keys()
repositories = set(r1) | set(r2)
for repo in repositories:
    # なんか処理
```

このとき片方はエラーで存在しない場合が想定されるのでちゃんとエラーハンドリングをして，存在しない場合にどういう値を入れるかを考えましょう．

次に辞書をマージする方法ですが，ある辞書d1とある辞書d2があったとき，Python 3.5以降ならマージした新しい辞書を以下の様にして生成できます．

```python
merged_dict = {**d1, **d2}
```

d1とd2で辞書のキーが被っている場合はd2のキーで上書きされます．後はこれを代入すれば良いでしょう

```python
self.all_results['repository'][dict_key] = merged_dict
```

pddg commented:
これはぼくの流儀なのでタスク次第では全く必要なく無駄なのですが，基本的に外部のAPIを叩いて取れる値は生のまま全て保存するようにしています．すると途中で集計方法が間違っていると気付いても，再リクエストすることなくローカルで完結できるからです．Issueのデータもとりあえず全部生のままJSONに突っ込んでおくと，後で必要なときに再利用が効く……かもしれません（ラベルの値やその作成時間が必要で無いなら不要かな？）．そのときにJSONが死ぬほどでかくなる（> 1GB以上）なら，分割して保存することを検討しましょう．

あと出力するJSONはindent入れずに全て1行で吐かせた方が小さいサイズのJSONになりますよ．

LakeRainSound commented:
@pddg 
今repositoryのmerge部分に手をつけています．setで和集合をとったときに，順序がバラバラになるため，出てくるjsonの順序もばらけてしまうのですが，これはどうしようもないのでしょうか？最悪，入力のファイルが順序を持っているので，そこから順序を特定はできますが...
```
import json

dict1 = {'repository': {
                    "a": "aaa"
                }
         }
dict2 = {'repository': {
                    "b": "bbb"
                }
         }

dict_list = [dict1, dict2]

r1 = dict1['repository'].keys()
r2 = dict2['repository'].keys()

# リポジトリのキーが全て入っている状態
repositories = set(r1) | set(r2)

all_result = {'repository': {}}
for repo_name in repositories:
    print(repo_name)

    merge_dict = {}
    for dict_ in dict_list:
        if repo_name in dict_['repository']:
            merge_dict = {**merge_dict, **dict_['repository']}

    all_result["repository"][repo_name] = merge_dict

print(json.dumps(all_result, indent=4))
```
このコードの結果は以下です．
```
b
a
{
    "repository": {
        "b": {
            "b": "bbb"
        },
        "a": {
            "a": "aaa"
        }
    }
}
```

pddg commented:
そもそもJSONのハッシュマップの読み込み順序は不定だと思うので，そこは順序を気にしなくても良いのでは？というのは少し思う（処理系依存な気がする）．どうしても一致させたい場合，現状では難しいので，最初に読み取ったリポジトリのリスト（ https://github.com/LakeRainSound/get_code_metrics/blob/master/get_code_metrics/main.py#L15 ）をそのままキーとして使用するのはどうでしょうか．現状だと，clocもGitHub APIの結果もエラーであったものは結果から消えますが，これを使えばどちらもエラーであったものも処理できるのでは？

LakeRainSound commented:
@pddg 
わかりました．入力ファイルが順序をもっているので，gcmでは結果の順序は不定というのを仕様ということにしておきます．結果を処理する時は入力ファイルを使って，処理するという形にします．
次に，実装マージのコードを実装しました．
```
def _get_linked_github_api_result(result_list):
    if len(result_list) == 0 or None in result_list:
        return None

    repositories = {}
    # key(repository name)を全て取得
    for result in result_list:
        repositories = set(repositories) | set(result.keys())

    github_results = {'repository': {}}
    for repo_name in repositories:
        # まずはgithub_resultsに存在しないキー(repo name)を追加
        github_results['repository'].setdefault(repo_name, {})
        for result in result_list:
            # resultにrepo_nameをキーとする情報が含まれているならgithub_resultsにマージ
            if repo_name in result:
                github_results['repository'][repo_name] = \
                                                    {**github_results['repository'][repo_name],
                                                     **result[repo_name]}
    return github_results
```
これで問題がなければoutputの方でも同様に記述したいと思うのですがどうでしょうか．

pddg commented:
無駄にディクショナリをネストさせる癖が付いているような気がします．そのせいで可読性が著しく下がっています．

```python
github_results = {}
for repo_name in repositories:
    merged = {}
    for result in result_list:
        current_res = result.get(repo_name)
        # 値が無い時はcontinue
        # ほんとは何が原因で値が無いのかとかを入れた方が親切
        if current_res is None:
            continue
        merged = {**merged, **current_res}
    github_results[repo_name] = merged
return {'repository': github_results}
```

pddg commented:
（あとガンガンこのブランチへpushして良いですよ．するとコードレビューの機能でソースコードに直接コメントできるので）

LakeRainSound commented:
@pddg 
今のコードでは存在しないrepositoryの結果も出力されることになるのですが，以下の例のようにエラーがあることを伝えるような要素を追加した方がいいですか？もしくは解析する時にキーが存在せず弾くことができるので必要ありませんか？
```
"LakeRainSound/empty": {
            "hasErrorLabelMetrics": true,
            "cloc": {
                "hasFile": false
            }
        }
```

pddg commented:
オススメはエラー原因を全て記録することです．実験結果を論文に書くときに，失敗した原因も一緒に書くことが多いので．

LakeRainSound commented:
@pddg 
エラーが存在するなら，"error"キーを作るといった形でいいでしょうか？
つまり，
```
"LakeRainSound/empty": {
            "repository": {
                "errors": [
                    {
                      "type": "NOT_FOUND",
                      "path": [
                        "repository"
                      ],
                      "locations": [
                        {
                          "line": 2,
                          "column": 3
                        }
                      ],
                      "message": "Could not resolve to a User with the username 'LakeRainSoundsound'."
                    }
                ]
            },
            "cloc": {
                "error": [
                    "Unable to Read:",
                    "hoge"
                ],
                "hasFile": false
            }
        }
```
のようにする形です．
エラーが存在しないならエラーキーは生成されません．

pddg commented:
形式は任せます．僕はどういう形式でエラーが出てくるのかは詳しくないので．後から見て，それを自分がちゃんと判別できる形式ならそれで良いと思います．

LakeRainSound commented:
@pddg 
この形式でいくことにします．
```
"lslakaflkjc/vhjksfhkavvn": {
            "errors": [
                {
                    "type": "NOT_FOUND",
                    "path": [
                        "repository"
                    ],
                    "locations": [
                        {
                            "line": 3,
                            "column": 23
                        }
                    ],
                    "message": "Could not resolve to a User with the username 'lslakaflkjc'."
                }
            ],
            "cloc": {
                "error": "\n1 error:\nUnable to read:  /Users/restartsugar/.cache/gcm/repositories/github.com/lslakaflkjc/vhjksfhkavvn\n"
            }
        }
```

LakeRainSound commented:
@pddg 
post_queryに手をつけています．
リトライは以下を参考にすればよいでしょうか．
https://qiita.com/toshitanian/items/c28a65fe2f32884e067c

次に，リトライしてもエラーだった場合ですが，処理は止めたくないので，エラーメッセージは表示しつつ，存在しなかった場合にはエラー文をjsonに記録して続行したいと思っています．
この場合，例えばrepositoryの情報を取得する場合，github_api/repository_info.pyにおいてget_all_repositories_infoのfor文の中にtry, except文を書けばよいのでしょうか？またexceptでとりあげるエラーは`except Exception`で記述すれば問題ないでしょうか．
```
import get_code_metrics.github_api.post_query as post_query


class RepositoryInfo:
    def __init__(self, token):
        self.access_token = token

    @staticmethod
    def create_repository_info_query(name_with_owner: str):
        query_state = """
        query{
          repository(owner: "%s", name: "%s") {
            nameWithOwner
            createdAt
            stargazers {
              totalCount
            }
            hasIssuesEnabled
            isArchived
            isFork
            isDisabled
            url
          }
          rateLimit {
            remaining
          }
        }
        """

        # owner, nameが存在しない場合はNoneをreturn
        if '/' in name_with_owner:
            owner = name_with_owner.split('/')[0]
            name = name_with_owner.split('/')[1]
            query_state = query_state % (owner, name)
        else:
            return None

        # queryとして送出できる形にする
        res_query = {'query': query_state}
        return res_query

    def get_repository_info(self, name_with_owner):
        # queryを作成
        query = self.create_repository_info_query(name_with_owner)
        # queryとアクセストークンを渡してpost
        repository_info = post_query.post_query(query, self.access_token)

        # repositoryが存在しないならerrorオブジェクトを返す
        if 'errors' in repository_info:
            return {'errors': repository_info['errors']}

        # API制限回避のためrateLimitが1000以下ならsleep
        post_query.avoid_api_limit(repository_info)

        return repository_info['data']['repository']

    def get_all_repositories_info(self, repository_list):
        res_all_repository = {}

        # APIがはじめに制限にかかりそうならsleepを挟む
        post_query.first_avoid_api_limit(self.access_token)

        for repository in repository_list:
            repository_info = self.get_repository_info(repository)

            res_all_repository.update({repository: repository_info})

        return res_all_repository
```

pddg commented:
良いと思います．これは実験用なので，とりあえずあらゆるExceptionをキャッチしても問題ないと思います．

LakeRainSound commented:
@pddg 
HTTPSリクエストのエラーハンドリングをしたコードを書きました．問題点があればしてきしていただきたいです．
変更したコードは，github_api中のコードになります．

pddg commented:
```python
try:
  # Some lines
except Exception:
  raise
```

という記述は

```python
# Some lines
```

と同義です

pddg commented:
（今更だけど，僕がコラボレータに入らないとコードレビューは使えないんでした…）

pddg commented:
スタックトレースの取得はsysモジュールよりもtracebackモジュールを使った方が楽かも知れません．

```python
import traceback

try:
    # Some lines
except Exception:
    # 文字列でスタックトレースを取得
    tb = traceback.format_exc()
```

pddg commented on get_code_metrics/code_metrics/cloc/cloc_analyze.py:
```diff
@@ -54,13 +55,14 @@ def get_cloc_results(self):
             # clocの結果をdict化する
             cloc_json = self.get_cloc_dict(repository_name, cloc_result_list)
-            res.update(cloc_json)
+            cloc_result.update(cloc_json)
```
`_get_error_message`も`get_cloc_dict`もどちらも同じような構造を返していて無駄なので，ここもう少し賢くしたいですね．

LakeRainSound replied:
@pddg 
hogeだけ返して，返ってきた部分でrepository_name: { "cloc": {hoge}}を追加する形でいいですかね？

pddg replied:
かなぁ．ただそれだと結局2箇所に`repository_name: { "cloc": {hoge}}`を書くことになるので，まずは代入する値の生成（エラーでも正常でも同じ変数名）→値の代入って感じにしたほうがいいかも．

LakeRainSound replied:
@pddg 
変更してみました．

pddg commented on get_code_metrics/code_metrics/cloc/cloc_analyze.py:
```diff
@@ -57,12 +57,9 @@ def get_cloc_results(self):
+                cloc_json = self._get_error_massage(cloc_result_list)
+            else:
+                cloc_json = self.get_cloc_dict(cloc_result_list)
+            cloc_result.update({repository_name: {"cloc": cloc_json}})
```
これ挙動としてかなり分かりづらいので，`cloc_result[repository_name] = {"cloc": cloc_json}`の方がいいと思います

LakeRainSound replied:
了解です．

LakeRainSound commented:
@pddg 
tqdmを導入しました．
loggingも使っているのですがディレクトリを用意してlogファイルの出力はあった方がいいですか？

pddg commented on get_code_metrics/github_api/label_info.py:
```diff
@@ -105,6 +110,12 @@ def get_all_repositories_label_metrics(self, repository_list):
+                    unit="repo")
+
+        logger.info('GitHub API(Label Info) Start')
         for repository in repository_list:
```
これは単純なリストなので，`for repository in tqdm(repository_list, desc=..., unit=...):`だけで良いです．totalを渡す必要は有りません．

LakeRainSound replied:
わかりました．

pddg commented:
loggingは一旦このPRでの導入はやめておきましょう．今すぐprintでは満たせない要件があるなら別ですが，全体でloggingに乗り換えないと意味が無いので．

LakeRainSound commented:
@pddg 
変更しました．
あとは気になっていることなんですが，setup.cfgに何を記述したらいいのか(特に依存等)がよくわかっていないです．

pddg commented:
setup.cfgのinstall_requiresは，このPythonモジュールが動作するために依存しているサードパーティーライブラリを列挙します．なので，今回で言うとrequestsとtqdmですね．
